src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransLiterals.java

```diff
@@ -0,0 +1,639 @@
+                JCIdent ident = makeParamIdent(params, params.head.name);
+                JCAssign assign = make.Assign(select, createApplyToList(List.of(ident), null));
+                assign.type = ident.type;
+                method.addStatement(make.Exec(assign));
+            } else {
+                for (JCVariableDecl field : fields) {
+                    JCFieldAccess select = makeThisFieldSelect(stringTemplateClass.type, field);
+                    JCIdent ident = makeParamIdent(params, field.name);
+                    JCAssign assign = make.Assign(select, ident);
+                    assign.type = ident.type;
+                    method.addStatement(make.Exec(assign));
+                }
+            }
+        }
+
+        void createValuesMethod() {
+            Type objectListType = makeListType(syms.objectType);
+            MethodInfo method = createMethod(SYNTHETIC | PUBLIC, names.values,
+                    objectListType, List.nil(), stringTemplateClass);
+            method.addStatement(make.Return(createApplyToList(createAccessors(), syms.objectType)));
+        }
+
+        void createInterpolateMethod() {
+            interpolateMethod = createMethod(SYNTHETIC | PUBLIC, names.interpolate,
+                    syms.stringType, List.nil(), stringTemplateClass);
+            interpolateMethod.addStatement(make.Return(concatExpression(fragments, createAccessors())));
+        }
+
+        void createToStringMethod() {
+            MethodInfo toStringMethod = createMethod(PUBLIC, names.toString,
+                    syms.stringType, List.nil(), stringTemplateClass);
+            JCExpression applytoString = this.createApply(syms.stringTemplateType, names.toString,
+                    List.of(make.This(stringTemplateClass.type)));
+            toStringMethod.addStatement(make.Return(applytoString));
+        }
+
+        void createHashCodeMethod() {
+            MethodInfo toHashCodeMethod = createMethod(PUBLIC, names.hashCode,
+                    syms.intType, List.nil(), stringTemplateClass);
+            JCExpression applyHashCode = this.createApply(syms.stringTemplateType, names.hashCode,
+                    List.of(make.This(stringTemplateClass.type)));
+            toHashCodeMethod.addStatement(make.Return(applyHashCode));
+        }
+
+        void createEqualsMethod() {
+            MethodInfo toEqualsMethod = createMethod(PUBLIC, names.equals,
+                    syms.booleanType, List.of(syms.objectType), stringTemplateClass);
+            List<JCVariableDecl> params = toEqualsMethod.decl.params;
+            JCIdent ident = makeParamIdent(params, params.head.name);
+            JCExpression applyEquals = this.createApply(syms.stringTemplateType, names.equals,
+                    List.of(make.TypeCast(syms.objectType, make.This(stringTemplateClass.type)), ident));
+            toEqualsMethod.addStatement(make.Return(applyEquals));
+        }
+
+        private JCClassDecl newStringTemplateClass() {
```

Comment:
The main consideration is performance. I spent quite a bit of time playing around with different implementations including metafactories (hence the carrier class work.) Since a majority of use cases will be STR and FMT, the number of classes will likely be just a few per application. Because of the change to force processor always, I will be revisiting this during the preview period to work on other solutions (I mentioned ProcessorFactories/ProcessorBuilders earlier).

Reply:
@JimLaskey, someone will implement a LOG at some point in the future and you will get many template classes per application class.
You mention the carrier but i believe you can implement something similar to a carrier erasing the type of the values but without using method handles to try to make it type safe and instead pass the real types as a class data of a hidden class and later as a type argument once the reified generics will be released.